experimental/overhead_matching/swag/data/vigor_dataset.py

```diff
@@ -98,6 +102,51 @@ def __getitem__(self, idx):
     def __len__(self):
         return len(self._panorama_metadata)
     
+    def generate_random_path(self,
+                             generator: torch.Generator,
+                             max_length_m: float,
+                             turn_temperature: float)->list[int]:
+        path = [] 
+        distance_traveled_m = 0.0
+        last_direction = torch.zeros(2)
+        # select start panorama
+        last_index = torch.randint(0, len(self), (1,), generator=generator)
+
+        def softmax_w_temp(x: torch.Tensor, t: float)-> torch.tensor:
+            ex = torch.exp(x / t)
+            return ex / torch.sum(ex)
+
+        while distance_traveled_m < max_length_m:
+            path.append(last_index)
+            last_item = self[last_index]
+            last_coord = torch.tensor([last_item.panorama_metadata['lon'], last_item.panorama_metadata['lat']])
+            last_neighbors = last_item.panorama_metadata['neighbor_panorama_idxs']
+
+            # calculate distribution over neighbors
+            neighbor_probs = torch.zeros(len(last_neighbors))
+            neighbor_coords = torch.zeros(len(last_neighbors), 2)
+            neighbor_directions = torch.zeros(len(last_neighbors), 2)
+            for i, neighbor_pano_idx in enumerate(last_neighbors):
+                neighbor_item = self[neighbor_pano_idx]
+                neighbor_coords[i] = torch.tensor([neighbor_item.panorama_metadata['lon'], neighbor_item.panorama_metadata['lat']])
+                neighbor_directions[i] = neighbor_coords[i] - last_coord
+                similarity = torch.dot(neighbor_directions[i], last_direction) / torch.norm(neighbor_directions[i]) 
+                if torch.is_nonzero(torch.norm(last_direction)):
```

Comment:
when would it not be non zero? On the first iteration? consider having some default initial direction and removing the if statement.